### ISSUE TITLE
 Can't resolve 'fs' and  Can't resolve 'path'

### ISSUE THREAD

**Issue by alishahbaz659** [electron-react-boilerplate/electron-react-boilerplate#3340]
When I try to use fs and path modules I am getting the below error: Please help 

ERROR in ./src/renderer/components/Camera.tsx 9:0-20
Module not found: Error: Can't resolve 'fs' in 'electron-react-boilerplate-main\src\renderer\components'
 @ ./src/renderer/App.tsx 9:0-41 46:163-169
 @ ./src/renderer/index.tsx 5:0-24 8:17-20

ERROR in ./src/renderer/components/Camera.tsx 10:0-24
Module not found: Error: Can't resolve 'path' in \electron-react-boilerplate-main\src\renderer\components'

BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
This is no longer the case. Verify if you need this module and configure a polyfill for it.

If you want to include a polyfill, you need to:
        - add a fallback 'resolve.fallback: { "path": require.resolve("path-browserify") }'
        - install 'path-browserify'
If you don't want to include a polyfill, you can use an empty module like this:
        resolve.fallback: { "path": false }
 @ ./src/renderer/App.tsx 9:0-41 46:163-169
 @ ./src/renderer/index.tsx 5:0-24 8:17-20

**Comment by RPruvost** [comment]
Hello 👋

I ran into the same issue, fixed it by adding this to .erb/webpack.config.base.ts :

```  
resolve: {
    fallback: {
      fs: false,
      tls: false,
      net: false,
      path: false,
      zlib: false,
      http: false,
      https: false,
      stream: false,
      crypto: false,
    },
    extensions: ['.js', '.jsx', '.json', '.ts', '.tsx'],
    modules: [webpackPaths.srcPath, 'node_modules'],
  },
  ```
  
  
  More information on this thread : https://stackoverflow.com/a/64802753
  Hope that will help.

**Comment by angelinimattia** [comment]
More precisely they blocked the system library for security reasons in the new versions of Electron.
1. you can unblock them as @RPruvost suggests ( I didn't try this  way)
2. You have to implement the code outside the web app (src/main) and use the IPC (inter-process communication) to request the main application to execute the related commands. Still, it is very tedious and prone to async-related problems.

**Comment by nsrCodes** [comment]
@RPruvost 's suggestion helps remove the original error from the console but if I try to add a simple IPC handler inside the UI as such

```diff
+import { ipcRenderer } from 'electron';
+import { useEffect } from 'react';
 import { MemoryRouter as Router, Routes, Route } from 'react-router-dom';
 import icon from '../../assets/icon.svg';
 import './App.css';
 
 const Hello = () => {
+  useEffect(() => {
+    ipcRenderer.on('example-event', (_event, payload) => {
+      // eslint-disable-next-line no-console
+      console.log(payload);
+    });
+  }, []);
+
   return (
     <div>
       <div className="Hello">
```


it gives the following error
```error
Uncaught ReferenceError: __dirname is not defined
    at ./node_modules/electron/index.js (index.js:4:1)
    at options.factory (react refresh:6:1)
    at __webpack_require__ (bootstrap:24:1)
    at fn (hot module replacement:62:1)
    at ./src/renderer/App.tsx (styleTagTransform.js:16:1)
    at options.factory (react refresh:6:1)
    at __webpack_require__ (bootstrap:24:1)
    at fn (hot module replacement:62:1)
    at ./src/renderer/index.tsx (App.tsx:59:2)
    at options.factory (react refresh:6:1)
```

I am not sure if this is a separate issue or an extension of the same.
_P.S. I have made all these changes in a fresh clone, and these are the only two changes that I have made_

**Comment by dmatora** [comment]
@RPruvost I tried that and now it crashes with
```
__dirname is not defined
```

**Comment by RPruvost** [comment]
> @RPruvost I tried that and now it crashes with
> 
> ```
> __dirname is not defined
> ```

In your main file (src/main/main.ts), do you have this ?
```js
    webPreferences: {
      nodeIntegration: true
      }
      ```
    __dirname is a Node.js global, you cannot use it without access to the Node.js APIs:
    https://nodejs.org/docs/latest/api/globals.html#globals_dirname

**Comment by subatomicglue** [comment]
@RPruvost  I tried your `resolve: { fallback: {...` code above, and now I get `__dirname is not defined`.  
However, adding `nodeIntegration: true` in `main.ts` did nothing, I still get the `__dirname is not defined`.

In my case: 
I am trying to `import { clipboard } from 'electron'` in the renderer in [`App.tsx`](https://github.com/subatomicglue/flaming-monkey-head-musicplayer-react/blob/main/src/renderer/App.tsx).
The moment that I `let cb = clipboard;` then I get the dreaded `Can't resolve 'fs' in '/path/to/node_modules/electron'`.
And then when I add the `fallback` code to [`.erb/config/webpack.config.base.ts`](https://github.com/subatomicglue/flaming-monkey-head-musicplayer-react/blob/main/.erb/configs/webpack.config.base.ts), I then get `__dirname is not defined`.

Ideas?
Code's [here](https://github.com/subatomicglue/flaming-monkey-head-musicplayer-react),

**Comment by subatomicglue** [comment]
same 

I got past my own issue (using the `clipboard`) by avoiding it (access the `clipboard` in `main` instead).   Electron Clipboard documentation makes it seem like it should work in both renderer and in main, but I found that not to be true, at least while I had `ipc` in the renderer.  So,  For future people, I got `electron` `clipboard` to work by calling it using `ipcRenderer` from the `renderer` to `main`'s process.    So in `main.tsx` I added `import { clipboard } from 'electron';` and a handler to expose the `clipboard`.   I _suspect_ that having the `ipcRenderer` in the `renderer` was confounding my effort above.

**Comment by alishahbaz659** [comment]
Doing the necessary IPC communication solved the issue for me. I implemented functionality that requires fs modules on the main thread and sends the response back to the renderer process. It is completely restricted in the latest version of the electron react boilerplate. Thank you for your help.

**Comment by wjyzhixing** [comment]
> Doing the necessary IPC communication solved the issue for me. I implemented functionality that requires fs modules on the main thread and sends the response back to the renderer process. It is completely restricted in the latest version of the electron react boilerplate. Thank you for your help.

I want to know how you do to resolve it?

**Comment by angelinimattia** [comment]
> > Doing the necessary IPC communication solved the issue for me. I implemented functionality that requires fs modules on the main thread and sends the response back to the rendered process. It is completely restricted in the latest version of the electron react boilerplate. Thank you for your help.
> 
> I want to know how you do to resolve it?

There is a section in the docs if you want an example [https://www.electronjs.org/docs/latest/tutorial/ipc](https://www.electronjs.org/docs/latest/tutorial/ipc)

That said...
First you have to understand the component of Electron(the name are not exact but is more to be able to talk):

1. Main Thread (that run the browser and interface with the system)
2. PreLoad (security layer to handle the communication between WebApp and main thread)
3. Web Application (the be app with your UI)

System related libraries (libraries that works with node.js for example) can only be loaded in the main thread, and not in the WebApp. The PreLoad is simply an isolation techniques to improve security, that manage the communication of the WebApp with the Main Thread.

To use the IPC(inter-Process Comunication) you need to do 3 things send the request (like you are talking with a server), ass the request to the preload (only to restrict the scope for security) handle the request on the main thread. 

SO let's take the example of the docs to sent a request to change the name showed on the browser window:
**1: Create the call in the preload to enable the front-end to talk with the backed:
```js
const { contextBridge, ipcRenderer } = require('electron')
 
// we tell the context bridge that we want to register a function undet the browser window key of ElectronAPI,
// a function callable from the browser called **setTitle**  that will send to the main process a request with ID _'set-title'_ and payload _'title'_
// Note that you can change the ElectronAPI with what you want but then you need to use window.your_choice.function, when you use it in the webApp!
contextBridge.exposeInMainWorld('electronAPI', {
  setTitle: (title) => ipcRenderer.send('set-title', title) 
})
```

**2: Now we can call this exposed function from the WebApp, I'm using here a  simple js script that we can load in the main.html, but you can use it in angular or react components too, since the function is registered in the browser's window object :**
```js
const setButton = document.getElementById('btn')
const titleInput = document.getElementById('title')
setButton.addEventListener('click', () => {
  const title = titleInput.value
  // Call the function registered in the window under 'ElectronAPI', as said you can change this by chagning the preload file 
  // if you want to use something different from 'ElectronAPI'
  window.electronAPI.setTitle(title) 
})
```

**3: Main thread handle the request:**
```js
const { app, BrowserWindow, ipcMain } = require('electron')
const path = require('node:path')

function createWindow () {
  const mainWindow = new BrowserWindow({
    webPreferences: {
      preload: path.join(__dirname, 'preload.js')
    }
  })
  // IPC request handling the IPC message is composed by a request ID('set-title') and a payload if needed (here is the title)
  // you need to register it in the window creation
  ipcMain.on('set-title', (event, title) => {
    const webContents = event.sender
    const win = BrowserWindow.fromWebContents(webContents)
    win.setTitle(title)
  })

  mainWindow.loadFile('index.html')
}

app.whenReady().then(() => {
  createWindow()

  app.on('activate', function () {
    if (BrowserWindow.getAllWindows().length === 0) createWindow()
  })
})

app.on('window-all-closed', function () {
  if (process.platform !== 'darwin') app.quit()
})
```

Hope it helps!

**Comment by wjyzhixing** [comment]
Because of your opition,.I resolve my problem .Thanks you very much.

**Comment by wjyzhixing** [comment]
> > > Doing the necessary IPC communication solved the issue for me. I implemented functionality that requires fs modules on the main thread and sends the response back to the rendered process. It is completely restricted in the latest version of the electron react boilerplate. Thank you for your help.
> > 
> > 
> > I want to know how you do to resolve it?
> 
> There is a section in the docs if you want an example https://www.electronjs.org/docs/latest/tutorial/ipc
> 
> That said... First you have to understand the components of Electron(the name are not exact but is more to be able to talk):
> 
> 1. Main Thread (that run the browser and interface with the system)
> 2. PreLoad (security layer to handle the communication between WebApp and main thread)
> 3. Web Application (the be app with your UI)
> 
> System related libraries (libraries that works with node.js for example) can only be loaded in the main thread, and not in the WebApp. The PreLoad is simply an isolation techniques to improve security, that manage the communication of the WebApp with the Main Thread.
> 
> To use the IPC(inter-Process Comunication) you need to do 3 things send the request (like you are talking with a server), ass the request to the preload (only to restrict the scope for security) handle the request on the main thread.
> 
> SO let's take the example of the docs to sent a request to change the name showed on the browser window: **1: Create the call in the preload to enable the front-end to talk with the backed:
> 
> ```js
> const { contextBridge, ipcRenderer } = require('electron')
>  
> // we tell the context bridge that we want to register a function under the browser window key of ElectronAPI,
> // this function will be callable from the browser, in this case the function is called **setTitle**  that will send to the main process a request with ID _'set-title'_ and payload _'title'_
> // Note that you can change the ElectronAPI with what you want but then you need to use window.your_choice.function, when you use it in the webApp!
> contextBridge.exposeInMainWorld('electronAPI', {
>   setTitle: (title) => ipcRenderer.send('set-title', title) 
> })
> ```
> 
> **2: Now we can call this exposed function from the WebApp, I'm using here a simple js script that we can load in the main.html, but you can use it in angular or react components too, since the function is registered in the browser's window object :**
> 
> ```js
> const setButton = document.getElementById('btn')
> const titleInput = document.getElementById('title')
> setButton.addEventListener('click', () => {
>   const title = titleInput.value
>   // Call the function registered in the window under 'ElectronAPI', as said you can change this by chagning the preload file 
>   // if you want to use something different from 'ElectronAPI'
>   window.electronAPI.setTitle(title) 
> })
> ```
> 
> **3: Main thread handle the request:**
> 
> ```js
> const { app, BrowserWindow, ipcMain } = require('electron')
> const path = require('node:path')
> 
> function createWindow () {
>   const mainWindow = new BrowserWindow({
>     webPreferences: {
>       preload: path.join(__dirname, 'preload.js')
>     }
>   })
>   // IPC request handling the IPC message is composed by a request ID('set-title') and a payload if needed (here is the title)
>   // you need to register it in the window creation
>   ipcMain.on('set-title', (event, title) => {
>     const webContents = event.sender
>     const win = BrowserWindow.fromWebContents(webContents)
>     win.setTitle(title)
>   })
> 
>   mainWindow.loadFile('index.html')
> }
> 
> app.whenReady().then(() => {
>   createWindow()
> 
>   app.on('activate', function () {
>     if (BrowserWindow.getAllWindows().length === 0) createWindow()
>   })
> })
> 
> app.on('window-all-closed', function () {
>   if (process.platform !== 'darwin') app.quit()
> })
> ```
> 
> Hope it helps!

extremely grateful！

**Comment by MaGiiK02** [comment]
> MohamedElgharbawy

As I said 'electronAPI' is no more than a key, and you can use everything you want here, as you said thought this is dose not automatically work on ts and some typing code is needed to tell the compiler that Electron is extending the window object with extra fields.

**Comment by yaumi123** [comment]
If anyone needs this, here is  what works well in my local.
just a case for me and not promising it will fix your problem.

1. add handle function in your main.js like below:
   ``` 
   async function handleFileOpen(type, arg) {
        return await dialog(null, { properties:[], ...arg});
 }
 ipcMain.handle('dialog:showOpenDialog', async (event, arg) => {
     return await handleFileOpen('showOpenDialog', arg);
}
```
this is how you use node api/function which  you need in main process
2. expose them to renderer in preload.js using contextBridge:
```
const handler =  {
    dialog: {
       showOpenDialog:(options) => {
     return ipcRenderer.invoke('dialog:showOpenDialog', options);
}
  }
}
contextBridge.exposeInMainWorld('electron', handler);
export type ElectronHandler = typeof handler;
```
3. in your code to use it:
```
window.electron.dialog.showOpenDialog({properties: [xxx]}).then();
```

**Comment by shunmian** [comment]
@RPruvost, I have followed your suggestion to add 

```
resolve: {
    fallback: {
      fs: false,
      tls: false,
      net: false,
      path: false,
      zlib: false,
      http: false,
      https: false,
      stream: false,
      crypto: false,
    },
    extensions: ['.js', '.jsx', '.json', '.ts', '.tsx'],
    modules: [webpackPaths.srcPath, 'node_modules'],
  },
```
The `Can't resolve 'fs'` error is gone but when I `import fs from 'fs'; console.log(fs)` it shows `undefined`.
